### PR TITLE
`intervalToDuration` fix for end of month calculations

### LINKED ISSUE #2611: intervalToDuration is off by 1
calculating difference between two dates shows that `intervalToDuration` is off by 1:

```ts
const currTime = new Date();
const timeAhead = new Date(currTime);
timeAhead.setFullYear(currTime.getFullYear() + 1);
timeAhead.setMonth(currTime.getMonth() + 1);
      
const { years, months, days, hours, minutes } = intervalToDuration({
  start: currTime,
  end: aheadTime,
});
      
console.log(years, months, days, hours, minutes) 
// Outputs: 1 1 1 0 0, but expected: 1 1 0 0 0
```

Creating a new issue here since this was first reported in May 2020 but hasn't been resolved yet: https://github.com/date-fns/date-fns/issues/1778

### PR DESCRIPTION
**Pull request by fturmel**
Closes #2611, #1778, #2470

By using `add` internally instead of `sub`, the `intervalToDuration` function gives better results for end of month interval start dates while still passing all existing unit tests. It also means better interop with `add`, as discussed in #2611.

Before:
```js
intervalToDuration({start: new Date(2021, 7, 31), end: new Date(2021, 8, 30)})
// {years: 0, months: 0, days: 30, hours: 0, minutes: 0, seconds: 0}
```

With this PR:
```js
intervalToDuration({start: new Date(2021, 7, 31), end: new Date(2021, 8, 30)})
// {years: 0, months: 1, days: 0, hours: 0, minutes: 0, seconds: 0}
```

This also brings a few improvements:
- new RangeError thrown if interval start date is greater than end date. This is standard in all other functions handling intervals and per https://date-fns.org/docs/Interval
- additional unit tests covering new edge cases and thrown exceptions
- leaner implementation: removed dependencies to `compareAsc` and `isValid`, removed sign and Math.abs operations that were not required.

### CODE REVIEW

File: src/intervalToDuration/index.ts [hunk 1]
```diff
@@ -33,49 +33,36 @@ import sub from '../sub/index'
  * })
  * // => { years: 39, months: 2, days: 20, hours: 7, minutes: 5, seconds: 0 }
  */
-
-export default function intervalToDuration({ start, end }: Interval): Duration {
+export default function intervalToDuration(interval: Interval): Duration {
```
Comment (nsafai) [inline src/intervalToDuration/index.ts:36]: what's the benefit of changing the func signature here? can we avoid a breaking change to syntax?

File: src/intervalToDuration/index.ts [hunk 2]
```diff
@@ -33,49 +33,36 @@ import sub from '../sub/index'
...
   requiredArgs(1, arguments)
 
-  const dateLeft = toDate(start)
-  const dateRight = toDate(end)
+  const start = toDate(interval.start)
+  const end = toDate(interval.end)
```
Comment (nsafai) [inline src/intervalToDuration/index.ts:40]: if you really want to keep the internal variable names the same (while keeping the func signature the same as before), you could do something like this:
```suggestion
  start = toDate(start)
  end = toDate(end)
```

File: src/intervalToDuration/index.ts [hunk 3]
```diff
@@ -33,49 +33,36 @@ import sub from '../sub/index'
...
 
-  const duration = {
-    years: 0,
-    months: 0,
-    days: 0,
-    hours: 0,
-    minutes: 0,
-    seconds: 0,
+  const duration: Duration = {
+    years: differenceInYears(end, start),
```
Comment (nsafai) [inline src/intervalToDuration/index.ts:49]: nice cleanup here 🥇

File: src/intervalToDuration/index.ts [hunk 4]
```diff
@@ -33,49 +33,36 @@ import sub from '../sub/index'
...
   }
 
-  const sign = compareAsc(dateLeft, dateRight)
-
-  duration.years = Math.abs(differenceInYears(dateLeft, dateRight))
-
-  const remainingMonths = sub(dateLeft, { years: sign * duration.years })
-  duration.months = Math.abs(differenceInMonths(remainingMonths, dateRight))
+  const remainingMonths = add(start, { years: duration.years })
+  duration.months = differenceInMonths(end, remainingMonths)
```
Comment (nsafai) [inline src/intervalToDuration/index.ts:53]: this is a nice cleanup as well 👏

File: src/intervalToDuration/test.ts [hunk 5]
```diff
@@ -1,15 +1,16 @@
...
-describe('intervalToDuration', function () {
-  it('returns correct duration for arbitrary dates', function () {
+describe('intervalToDuration', () => {
+  it('returns correct duration for arbitrary dates', () => {
     const start = new Date(1929, 0, 15, 12, 0, 0)
     const end = new Date(1968, 3, 4, 19, 5, 0)
     const result = intervalToDuration({ start, end })
 
-    assert.deepEqual(result, {
+    assert.deepStrictEqual(result, {
```
Comment (nsafai) [inline src/intervalToDuration/test.ts:13]: even safer, nice

File: src/intervalToDuration/test.ts [hunk 6]
```diff
@@ -49,9 +50,36 @@ describe('intervalToDuration', function () {
...
+  })
+
+  it("throws a RangeError if interval's end date invalid", () => {
+    const interval = {
+      start: new Date(2020, 0, 1),
+      end: new Date(NaN),
+    }
+
+    assert.throws(intervalToDuration.bind(null, interval), RangeError)
+  })
```
Comment (nsafai) [inline src/intervalToDuration/test.ts:78]: nice additions

File: src/intervalToDuration/test.ts [hunk 7]
```diff
@@ -112,5 +140,147 @@ describe('intervalToDuration', function () {
...
+          months: 1,
+          days: 1,
+          hours: 0,
+          minutes: 0,
+          seconds: 0,
+        }
+
+        assert.deepStrictEqual(duration, expectedDuration)
+      })
+    })
```
Comment (nsafai) [inline src/intervalToDuration/test.ts:284]: nice these are all great edge cases to test

**Review by nsafai — COMMENTED** [review]
i think it's worth leaving the func signature as-is, other than that this looks great, tysm for following up on this!

**Comment by nsafai** [comment]
lmk what you think of my suggestion of keeping func signature as-is and feel free to tag me for a +1

File: src/intervalToDuration/index.ts (hunk 1 above)
Comment (fturmel) [reply src/intervalToDuration/index.ts:36]: Correct me if I'm wrong, but there are no signature changes here. It's not destructuring the object anymore but the only argument is still an Interval object. As I think you noticed, the change was to allow the use of `start` and `end` variable names in the function body for readability.

File: src/intervalToDuration/index.ts (hunk 2 above)
Comment (fturmel) [reply src/intervalToDuration/index.ts:40]: Personal preference, but I don't love assignments to function arguments. I tend to treat them as read only. In this case, they are also not quite the same type. Interval start and end are `Date | number` but the return value of `toDate` is `Date`.

**Comment by fturmel** [comment]
@nsafai Thanks for taking the time and all the nice notes! As I replied in my comments I think the method signature is intact but let me know if I'm missing something.

**Review by nsafai — APPROVED** [review]
ah you're entirely right, func signature is still backwards compatible. LGTM!

**Review by tan75 — APPROVED** [review]
Thank you!

**Comment by hiendaovinh** [comment]
Thanks for the fix but it's a breaking change because of the Math.abs removal. Please bump a major version next time.

**Comment by leshakoss** [comment]
@hiendaovinh sorry, that was not supposed to happen :[ A patch with a revert coming soon

**Comment by leshakoss** [comment]
@hiendaovinh actually, I don't think this is a breaking change. If interval.start was after interval.end the function would've thrown a RangeError anyway, so there's no way to get negative values from the function. If I'm wrong, please provide a test case that we can use to verify this 🙇

**Comment by fturmel** [comment]
>  If interval.start was after interval.end the function would've thrown a RangeError anyway

@leshakoss The RangeError was introduced in this PR, so if there are any breaking changes that would be it. However that error should've always been there if you look at the behavior of every other functions accepting an Interval in the library, per my PR notes.

**Comment by leshakoss** [comment]
Ah, I see. Honestly, this is a tough call. I would count this as a bug fix rather than a breaking change, but that's maybe because I already have "interval.start is not supposed to be after interval.end" in mind, but I understand why one might think the opposite

**Comment by hiendaovinh** [comment]
I do understand and I share the same thinking without regarding the "interval.start is not supposed to be after interval.end" stuff. Actually one of my colleges had a bug switching their places so I knew :laughing:. It's certainly a bug fix but also a breaking change in its own way. That' interesting.

**Comment by leshakoss** [comment]
@hiendaovinh reverted the change in v2.29.2: https://github.com/date-fns/date-fns/releases/tag/v2.29.2
